### PR TITLE
set dim[0] to -1 if dim[0] < 0 during compiling for c_allgather op

### PR DESCRIPTION
**Pull request by sandyhouse**
set dim[0] to -1 if dim[0] < 0 during compiling for c_allgather op
move the assertion (input.shape[0] must be divisible by nranks) from compiler time to run time

### CODE REVIEW

File: python/paddle/fluid/layers/collective.py [hunk 1]
```diff
@@ -134,10 +134,6 @@ def _c_reducescatter(x, nranks, ring_id=0, use_calc_stream=False):
     if not isinstance(x, Variable):
         raise TypeError('x must be a Variable')
 
-    if x.shape[0] % nranks != 0:
```
Comment (gavin1332) [inline python/paddle/fluid/layers/collective.py:137]: keep this check when shape[0] larger than 0

Comment (sandyhouse) [reply python/paddle/fluid/layers/collective.py:137]: done.

**Review by gavin1332 — DISMISSED** [review]
LGTM

**Review by gavin1332 — APPROVED** [review]
LGTM

**Pull request merged**